api/processors/services/group_builder.py

```diff
@@ -146,23 +150,120 @@ def _get_allowed_group_size(age_range: AgeRange) -> GroupSizeRestriction:
+            max_students_num=restrictions.max,
+        )
+        logger.debug(students)
+        if len(students) < restrictions.min:
+            return None
 
         return GroupCandidate(
-            age_range=AgeRange(age_from=18, age_to=90),
-            language_and_level=teacher.teaching_languages_and_levels.first(),  # type: ignore
-            communication_language_mode=CommunicationLanguageMode(
-                teacher.personal_info.communication_language_mode
-            ),
+            age_range=age_range,
+            language_and_level=language_and_level,
+            communication_language_mode=communication_language,
             teacher=teacher,
-            students=list(range(1, 10)),  # type: ignore
-            monday=timezone.now(),
-            wednesday=timezone.now(),
+            students=students,
+            day_time_slots=day_time_slots,
+        )
+
+    @staticmethod
+    def _get_students(
+        language_and_level: LanguageAndLevel,
+        age_range: AgeRange,
+        communication_language: CommunicationLanguageMode,
+        day_time_slots: Iterable[DayAndTimeSlot],
+        max_students_num: int,
+    ) -> Collection[Student]:
+        # TODO: handle age ranges properly (add more ranges when necessary)
+        student_manager = Student.objects.filter(
+            project_status=StudentProjectStatus.NO_GROUP_YET,
+            teaching_languages_and_levels=language_and_level,
+            personal_info__communication_language_mode=communication_language,
+            age_range__age_from__gte=age_range.age_from,
+            age_range__age_to__lte=age_range.age_to,
         )
+        for time_slot in day_time_slots:
+            student_manager = student_manager.filter(availability_slots=time_slot)
+
+        return student_manager.order_by("personal_info__date_and_time_added")[:max_students_num]
+
+    @staticmethod
+    def _iterate_lesson_times(teacher: Teacher) -> Iterator[tuple[DayAndTimeSlot, DayAndTimeSlot]]:
+        """
+        Iterates tuples of available time slots.
+        There should be at least one free day between lessons.
+        """
+        teacher_availability_slots = teacher.availability_slots.all()
+        for first_availability in teacher_availability_slots:
+            for second_availability in teacher_availability_slots:
+                if GroupBuilder._availability_slots_have_break(
+                    first_availability, second_availability
+                ):
+                    yield (first_availability, second_availability)
+
+    @staticmethod
+    def _availability_slots_have_break(
+        first_availability: DayAndTimeSlot, second_availability: DayAndTimeSlot
+    ) -> bool:
+        """
+        Checks that second availability slot is not too early from first
+        """
+        first_day = first_availability.day_of_week_index
+        second_day = second_availability.day_of_week_index
+        if second_day - first_day < MIN_DAYS_BETWEEN_LESSONS:
```

Comment:
I think either the constant must be `2` or it should be `<=` instead of `<` here
Also maybe we shouldn't count on the days being ordered and either sort them or compare `abs()`

Reply:
UPD see [this comment](https://github.com/samanthasgroup/django-webapps/pull/240#discussion_r1320614245) by @Roxe322 instead

Reply:
Good job spotting `<` instead of `<=`! As for `abs` - actually that was point to also filter out second <= first pairs. Added it to docstring to clarify